商品出品機能の実装

rinooooo opened the pull request:
#what
新規出品画面の作成、モデルの単体テスト
#why
商品出品機能を実装するため

よろしくお願いします！
・ログイン状態の場合は、商品出品ページへ遷移できる動画
https://gyazo.com/f037c7ad7b907886c02a1b16a8043511

・価格が入力されると同時に、販売手数料と販売利益が表示される動画
https://gyazo.com/f4998593b689f3f33ccac03fd3a55952

・必要な情報を適切に入力して「出品する」ボタンを押すと、商品情報がデータベースに保存される動画
https://gyazo.com/e4dd60a14e63b2f53710c9fa99fd1a94

・入力に問題がある状態で「出品する」ボタンが押された場合、情報は保存されず、出品ページに戻りエラーメッセージが表示される動画
https://gyazo.com/7a1f0f6ad66060a72c58e75acd735b99

・ログアウト状態の場合は、商品出品ページへ遷移しようとすると、ログインページへ遷移する動画
https://gyazo.com/284c6710243cdb0bbdf2ac7a2fc57a2b

 ・テスト結果の画像
https://gyazo.com/b4ad038a66e6f6142209f94d41bc5c38

rinooooo committed:
ルーティングの設定

rinooooo committed:
Itemモデルの作成

rinooooo committed:
itemsテーブルの作成

rinooooo committed:
Itemモデル、バリデーションの設定

rinooooo committed:
Itemモデル・Userモデルのアソシエーション

rinooooo committed:
itemsコントローラーの設定

rinooooo committed:
アクティブハッシュの設定(category))

rinooooo committed:
アクティブハッシュの設定(condition)

rinooooo committed:
アクティブハッシュの設定(delivery)

rinooooo committed:
ActiveStorageのインストール

rinooooo committed:
画像投稿機能の実装

rinooooo committed:
商品投稿機能の実装

rinooooo committed:
「出品する」のリンク条件分け

rinooooo committed:
エラーメッセージの表示

rinooooo committed:
item_priceのバリデーション設定

rinooooo committed:
販売手数料、販売利益の表示

rinooooo committed:
Itemモデルの単体テスト

rinooooo committed:
rubocopの実行

maimai0401 commented on app/controllers/items_controller.rb:
```diff
@@ -1,4 +1,24 @@
 class ItemsController < ApplicationController
   def index
   end
+
```
以下の実装条件を満たすための実装を行いましょう。
・ログイン状態のユーザーのみ商品出品ページへ遷移できること。

rinooooo replied:
修正しました。

rinooooo replied:
・ログアウト状態の場合は、商品出品ページへ遷移しようとすると、ログインページへ遷移する動画
https://gyazo.com/3d8f4b38c157a0e7b5b55e595d200cdb

maimai0401 commented on app/models/item.rb:
```diff
@@ -0,0 +1,25 @@
+  validates :delivery_area_id, numericality: { other_than: 1, message: "can't be blank" }
+  validates :delivery_day_id, numericality: { other_than: 1, message: "can't be blank" }
+
+  with_options presence: true, format: { with: /\A[0-9]+\z/, message: '半角数値を使用してください' } do
```
```
, format: { with: /\A[0-9]+\z/, message: '半角数値を使用してください' }
```
今回の実装において価格のカラム型がinteger型となっているため、上記記述は不要です。
削除しておきましょう。

rinooooo replied:
修正しました。

maimai0401 commented on app/views/items/new.html.erb:
```diff
@@ -5,10 +5,10 @@
-    <%= form_with local: true do |f| %>
+    <%= form_with model: @item, url: items_path, local: true do |f| %>
 
     <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
```
    <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
上記のようなコメントアウトは実装方針の内容であり、企業にポートフォリオとして提出する際に、好ましくないため削除しましょう。

rinooooo replied:
修正しました。

maimai0401 commented on app/views/items/new.html.erb:
```diff
@@ -5,10 +5,10 @@
     <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
-    <%# render 'shared/error_messages', model: f.object %>
+    <%= render 'shared/error_messages', user: @item %>
     <%# //インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
```
    <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
上記のようなコメントアウトは実装方針の内容であり、企業にポートフォリオとして提出する際に、好ましくないため削除しましょう。

rinooooo replied:
修正しました。

maimai0401 commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,133 @@
+RSpec.describe Item, type: :model do
+  before do
+    @item = FactoryBot.build(:item)
+    @item.image = fixture_file_upload('/hero.jpg')
```
```
    @item.image = fixture_file_upload('/hero.jpg')
```
画像の定義はspec/factories/items.rb内で行うようにしましょう。
理由は、FactoryBot内でデータの紐付けを行うことで、そのFactoryBotが他のモデルテストや結合テストでも使いまわしのしやすいデータになるためです。

rinooooo replied:
修正しました。

rinooooo replied:
・テスト結果の画像
https://gyazo.com/65cbbb09591348faecedd4faf22cfb33

maimai0401 reviewed (commented):
レビュー依頼ありがとうございます。
コメントいたしましたので、ご確認をお願いいたします。
修正後、再度レビュー依頼フォームにてご提出くださいませ。

rinooooo committed:
修正①

・コメントアウトの削除
・ログイン状態のユーザーのみ商品出品ページへ遷移する機能の書き方変更
・item_priceのバリデーション修正
・imageのバリデーション修正

rinooooo committed:
rubocopの実行

maimai0401 commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,120 @@
+require 'rails_helper'
+
```
下記のテストを追加で実装しましょう。
・価格に半角数字以外が含まれている場合は出品できない

rinooooo replied:
追加しました。
・テスト結果の画像
https://gyazo.com/496a937210269b86ed534fa93676e7d9

maimai0401 reviewed (commented):
レビュー依頼ありがとうございます。
修正点確認させていただきました。
コメントいたしましたので、ご確認お願いいたします。
修正後、再度レビュー依頼フォームにてご提出くださいませ。

rinooooo committed:
Itemモデルテストの追加

maimai0401 reviewed (commented):
レビュー依頼ありがとうございます。
LGTMとさせていただきます！実装お疲れ様でした。

マージ後、ローカルのmasterブランチに忘れずにプルしましょう。
その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。
引き続き実装頑張っていきましょう！
